packages/nx/src/project-graph/project-graph.ts

```
@@ -267,6 +275,35 @@ export async function createProjectGraphAndSourceMapsAsync(
   performance.mark('create-project-graph-async:start');
 
   if (!daemonClient.enabled()) {
+    const lock = new FileLock(
+      join(workspaceDataDirectory, 'project-graph.lock')
+    );
+
+    function cleanupFileLock() {
+      try {
+        lock.unlock();
+      } catch {}
+    }
+
+    process.on('exit', cleanupFileLock);
+
+    if (lock.locked) {
+      logger.verbose(
+        'Waiting for graph construction in another process to complete'
+      );
+      await lock.wait();
+      const sourceMaps = readSourceMapsCache();
+      if (!sourceMaps) {
+        throw new Error(
+          'The project graph was computed in another process, but the source maps are missing.'
+        );
+      }
+      return {
+        projectGraph: await readCachedGraphAndHydrateFileMap(),
```

This should throw an error if the graph is in a bad state